### PR TITLE
Rewrite Dockerfile to build the project

### PR DESCRIPTION
**Pull request by ahornace**
I've seen that it is quite common to build projects directly in `Dockerfile` so I thought I will give it a try. It is then very useful in `docker-compose` since you can build many things by a single command.

If the maven repositories are cached, then it takes about 6 minutes to build the whole docker image. But if travis does not use cache and builds it all from scratch, then it may significantly prolong the travis execution time which might not be that wise.

What do you guys think?

If you do not like it, then feel free to close it :)

### CODE REVIEW

**Comment by vladak** [comment]
> If the maven repositories are cached, then it takes about 6 minutes to build the whole docker image. But if travis does not use cache and builds it all from scratch, then it may significantly prolong the travis execution time which might not be that wise.

What actually happens now ? Is the caching in Travis builds used at all ?

File: Dockerfile [hunk 1]
```diff
@@ -1,34 +1,60 @@
-FROM debian:stable-slim as fetcher
+# Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
 
-# TODO copy just the 'distribution' directory, not all source code
+FROM ubuntu:bionic as build
```
Comment (vladak) [inline Dockerfile]: is this going to change the resulting OS in the image ?

Comment (ahornace) [reply Dockerfile]: No, it's just for the build

**Comment by ahornace** [comment]
> What actually happens now ? Is the caching in Travis builds used at all ?

I did not have time to check and play with it. However, is docker build even needed to be in travis? It is basically checking only `Dockerfile` correctness and takes a lot of time.

**Comment by vladak** [comment]
> > What actually happens now ? Is the caching in Travis builds used at all ?
> 
> I did not have time to check and play with it. However, is docker build even needed to be in travis? It is basically checking only `Dockerfile` correctness and takes a lot of time.

It was added to catch any Docker related problems early. Too late to do it during release build. Long running builds/tests are not a problem I'd say - as long as they do not hit the CI timeouts.

File: Dockerfile [hunk 2]
```diff
@@ -1,34 +1,60 @@
...
+RUN mkdir -p /mvn/opengrok-indexer/target/jflex-sources
+RUN mkdir -p /mvn/opengrok-web/src/main/webapp
+COPY opengrok-indexer/jflex-tt.txt /mvn/opengrok-indexer/jflex-tt.txt
+COPY opengrok-indexer/jflex-tt-end.txt /mvn/opengrok-indexer/jflex-tt-end.txt
+COPY opengrok-indexer/jflex-code.txt /mvn/opengrok-indexer/jflex-code.txt
+COPY opengrok-indexer/jflex-code-end.txt /mvn/opengrok-indexer/jflex-code-end.txt
+RUN mkdir -p /mvn/opengrok-web/src/main/webapp/WEB-INF/ && touch /mvn/opengrok-web/src/main/webapp/WEB-INF/web.xml
+
+# dummy build to cache the dependencies
+RUN mvn package -DskipTests -Dcheckstyle.skip -Dmaven.antrun.skip
```
Comment (vladak) [inline Dockerfile]:
```suggestion
RUN mvn -DskipTests -Dcheckstyle.skip -Dmaven.antrun.skip package
```

File: Dockerfile [hunk 3]
```diff
@@ -1,34 +1,60 @@
-FROM debian:stable-slim as fetcher
+# Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
 
-# TODO copy just the 'distribution' directory, not all source code
+FROM ubuntu:bionic as build
+
+RUN apt-get update && apt-get install -y openjdk-8-jdk python3 maven python3-venv
```
Comment (vladak) [inline Dockerfile]:
```suggestion
RUN apt-get update && apt-get install -y openjdk-8-jdk maven python3 python3-venv
```

**Comment by ahornace** [comment]
> It was added to catch any Docker related problems early. Too late to do it during release build. Long running builds/tests are not a problem I'd say - as long as they do not hit the CI timeouts.

That reminds me not to use `latest` tag which is a good practice. In that case there is basically nothing that can make the docker build to fail except change in the `mvn clean package` command which will probably stay the same for a long long time.

But probably better safe than sorry.

File: Dockerfile (hunk 2 above)
Comment (ahornace) [reply Dockerfile]: I like it the other way round but okay :D

File: Dockerfile (hunk 1 above)
Comment (vladak) [reply Dockerfile]: So how the resulting OS will get set ? (or where from)

Comment (ahornace) [reply Dockerfile]: From this command: `FROM tomcat:9-jre8`

File: Dockerfile [hunk 4]
```diff
@@ -1,34 +1,60 @@
...
+RUN sed -i 's:<module>distribution</module>::g' /mvn/pom.xml
+RUN sed -i 's:<module>opengrok-tools</module>::g' /mvn/pom.xml
+
+RUN mkdir -p /mvn/opengrok-indexer/target/jflex-sources
+RUN mkdir -p /mvn/opengrok-web/src/main/webapp
+COPY opengrok-indexer/jflex-tt.txt /mvn/opengrok-indexer/jflex-tt.txt
+COPY opengrok-indexer/jflex-tt-end.txt /mvn/opengrok-indexer/jflex-tt-end.txt
+COPY opengrok-indexer/jflex-code.txt /mvn/opengrok-indexer/jflex-code.txt
+COPY opengrok-indexer/jflex-code-end.txt /mvn/opengrok-indexer/jflex-code-end.txt
+RUN mkdir -p /mvn/opengrok-web/src/main/webapp/WEB-INF/ && touch /mvn/opengrok-web/src/main/webapp/WEB-INF/web.xml
```
Comment (vladak) [inline Dockerfile]: I think this section here actually makes it necessary to run Docker builds for every push/PR as it relies on internal structure of the project.

Comment (ahornace) [reply Dockerfile]: Yes, you are right :D I always forget how overcomplicated the build is :D

File: Dockerfile [hunk 5]
```diff
@@ -1,34 +1,60 @@
...
+WORKDIR /mvn
+COPY pom.xml /mvn/
+COPY opengrok-indexer/pom.xml /mvn/opengrok-indexer/
+COPY opengrok-web/pom.xml /mvn/opengrok-web/
+COPY plugins/pom.xml /mvn/plugins/
+COPY suggester/pom.xml /mvn/suggester/
+
+# distribution and opengrok-tools do not have dependencies to cache
+RUN sed -i 's:<module>distribution</module>::g' /mvn/pom.xml
+RUN sed -i 's:<module>opengrok-tools</module>::g' /mvn/pom.xml
```
Comment (vladak) [inline Dockerfile]: why is this necessary ?

File: Dockerfile (hunk 4 above)
Comment (vladak) [reply Dockerfile]: more than before I mean.

File: Dockerfile (hunk 5 above)
Comment (ahornace) [reply Dockerfile]: Because if they are present, then `mvn` call will try to execute the plugins in those submodules which will fail the build. And I did not want to create other "dummy" files for it to pass with no added value.

File: Dockerfile (hunk 2 above)
Comment (vladak) [reply Dockerfile]: I never liked the GNU style of arguments (mixing options with arguments). In this case I think the options are global and not specific to the `package` sub command therefore it makes sense to state them first.

**Comment by vladak** [comment]
> > It was added to catch any Docker related problems early. Too late to do it during release build. Long running builds/tests are not a problem I'd say - as long as they do not hit the CI timeouts.
> 
> That reminds me not to use `latest` tag which is a good practice. In that case there is basically nothing that can make the docker build to fail except change in the `mvn clean package` command which will probably stay the same for a long long time.

using the latest tag where ?

**Comment by vladak** [comment]
> I've seen that it is quite common to build projects directly in `Dockerfile` so I thought I will give it a try. It is then very useful in `docker-compose` since you can build many things by a single command.

could you give an example ? what do you do with docker compose and OpenGrok exactly ?

**Comment by ahornace** [comment]
> could you give an example ? what do you do with docker compose and OpenGrok exactly ?

At work I have a lot of projects that are being built by Dockerfile: 
- ui
- backend
- grafana with configuration

and other dependencies: db, kafka, prometheus

All I need to do is run `docker-compose build && docker-compose up` from any machine with docker and it just runs.

And at the moment I have OpenGrok with metrics being pushed to Graphite and shown in Grafana but it needs some final touches.

**Comment by ahornace** [comment]
> using the latest tag where ?

E.g. I should not use `ubuntu:bionic` because it seems it is being updated regularly but a specific tag which will stay the same forever, e.g. `ubuntu:bionic-20191029`.
